ipc/provider/src/manager/evm/dryrun.rs

```diff
@@ -0,0 +1,405 @@
+        self.caller.checkpoint_period(subnet_id).await
+    }
+
+    async fn checkpoint_bundle_at(
+        &self,
+        height: ChainEpoch,
+    ) -> anyhow::Result<BottomUpCheckpointBundle> {
+        self.caller.checkpoint_bundle_at(height).await
+    }
+
+    async fn quorum_reached_events(
+        &self,
+        height: ChainEpoch,
+    ) -> anyhow::Result<Vec<QuorumReachedEvent>> {
+        self.caller.quorum_reached_events(height).await
+    }
+
+    async fn current_epoch(&self) -> anyhow::Result<ChainEpoch> {
+        BottomUpCheckpointRelayer::current_epoch(&self.caller).await
+    }
+}
+
+#[async_trait]
+impl TopDownFinalityQuery for EvmSubnetDryRun {
+    async fn genesis_epoch(&self, subnet_id: &SubnetID) -> anyhow::Result<ChainEpoch> {
+        self.caller.genesis_epoch(subnet_id).await
+    }
+
+    async fn chain_head_height(&self) -> anyhow::Result<ChainEpoch> {
+        self.caller.chain_head_height().await
+    }
+
+    async fn get_top_down_msgs(
+        &self,
+        subnet_id: &SubnetID,
+        epoch: ChainEpoch,
+    ) -> anyhow::Result<TopDownQueryPayload<Vec<IpcEnvelope>>> {
+        self.caller.get_top_down_msgs(subnet_id, epoch).await
+    }
+
+    async fn get_block_hash(&self, height: ChainEpoch) -> anyhow::Result<GetBlockHashResult> {
+        self.caller.get_block_hash(height).await
+    }
+
+    async fn get_validator_changeset(
+        &self,
+        subnet_id: &SubnetID,
+        epoch: ChainEpoch,
+    ) -> anyhow::Result<TopDownQueryPayload<Vec<StakingChangeRequest>>> {
+        self.caller.get_validator_changeset(subnet_id, epoch).await
+    }
+
+    async fn latest_parent_finality(&self) -> anyhow::Result<ChainEpoch> {
+        self.caller.latest_parent_finality().await
+    }
+}
+
+impl EvmSubnetDryRun {
+    pub fn new(vm_type: VMType, caller: EthSubnetManager) -> Self {
+        Self { vm_type, caller }
+    }
+
+    /// A todo marker that does not panic
+    fn todo<T>(&self) -> anyhow::Result<T> {
+        Err(anyhow!("not implemented yet"))
+    }
+
+    fn create_subnet_evm(
+        &self,
+        from: &Address,
+        params: ConstructParams,
+    ) -> anyhow::Result<MockedEVMTxn> {
+        let converted = register_subnet_facet::ConstructorParams::try_from(params)?;
+        log::debug!("converted constructor params: {converted:?}");
+
+        let to = converted.ipc_gateway_addr;
+        let from = payload_to_evm_address(from.payload())?;
+        log::debug!("from: {}, to: {}", from, to);
+
+        let (calldata, method) = to_evm_calldata(
+            &register_subnet_facet::REGISTERSUBNETFACET_ABI.functions,
+            "newSubnetActor",
+            // ethers needs the params to be tuple here
+            (converted,),
+        )?;
+
+        Ok(MockedEVMTxn {
+            from,
+            to,
+            value: U256::zero(),
+            method,
+            calldata,
+        })
```

Comment:
This seems like a lot of code for the dry-run, which, AFAICS, will never actually be verified for correctness. That is, unlike what I suggested, which was:
1. Do everything normally, let the `ipc-cli` construct the _real_ calldata and send it to the API...
2. ... except in dry-run mode short-circuit before hitting the `eth_sendRawTransaction` endpoint and just print the parameters. 
3. This way we _know_ that what we print is correct, because we test it with _non-dry-run_ mode all the time. 

By contrast, here the dry-run mode is entirely separate not just in execution, but in its construction of the call parameters. It's nice that the ABI checks some of the compatibility things :crossed_fingers: but I'm not sure what would happen if e.g. we forgot to pass the argument as a tuple.

Reply:
Hopefully it's fine as `encode_function_data` does type checking, so for methods which have very different input this should be okay.